File: truss-compiler/src/main/java/truss/compiler/parser/IndexSelector.java
package truss.compiler.parser;

import truss.compiler.Span;
import truss.compiler.support.ImmutableArray;
import truss.compiler.syntax.ElementAccessExpressionSyntax;
import truss.compiler.syntax.ExpressionSyntax;

class IndexSelector implements Selector {
    private final ImmutableArray<ExpressionSyntax> expressions;
    private final Span span;

    public IndexSelector(ImmutableArray<ExpressionSyntax> expressions, Span span) {
        this.expressions = expressions;
        this.span = span;
    }

    @Override
    public ExpressionSyntax build(ExpressionSyntax value) {
        return new ElementAccessExpressionSyntax(value, expressions, span);
    }
}

File: Truss.Compiler/Parser/IndexSelector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Truss.Compiler.Support;
using Truss.Compiler.Syntax;

namespace Truss.Compiler.Parser {
    internal class IndexSelector : ISelector {
        private readonly ImmutableArray<ExpressionSyntax> _expressions;
        private readonly Span _span;

        public IndexSelector(ImmutableArray<ExpressionSyntax> expressions, Span span) {
            _expressions = expressions;
            _span = span;
        }

        public ExpressionSyntax Build(ExpressionSyntax value) {
            return new ElementAccessExpressionSyntax(value, _expressions, _span);
        }
    }
}
